rfd/0080-yubikey-piv.md

```diff
@@ -0,0 +1,272 @@
+---
+authors: Brian Joerger (bjoerger@goteleport.com)
+state: draft
+---
+
+# RFD 80 - yubikey PIV integration
+
+## What
+
+We will integrate yubikey smart card capabilities into the Teleport login process so that users can login with a yubikey generated private key. Teleport admins will be able to require yubikey login with optional PIN and touch policies for each cluster and role.
+
+## Why
+
+Currently, Teleport clients create their own public-private key pairs to be signed by the Teleport Auth server. The private key must be stored for future handshakes. In most cases, the private key is stored in disk alongside the certificates signed for it, where it can potentially exported.
+
+For example, `tsh login` stores the private key in `~/.tsh/keys/proxy/user` and stores certs in `~/.tsh/keys/proxy`, so a user's login session is fully contained in their `~/.tsh` folder. As a result, if someone gains access to a Teleport user's computer, they would immediately have access to the user's Teleport login session. 
+
+Using a hardware backed private key like yubikey, we can prevent such vulnerabilities. In order to gain access to the Teleport login session's private key, you would need to have access to the user's certificates on disk, their physical yubikey, and the ability to pass the yubikey's touch and pin policies.
+
+Additionally, Teleport currently has no way to determine the origins of a private key it signs, and simply trusts that it will be stored securely on the client side. With yubikey private keys, Teleport Auth servers will able to perform [Attestation](https://docs.yubico.com/yesdk/users-manual/application-piv/attestation.html) to confirm that a public key was securely generated in a yubikey before signing certificates for the key.
+
+## Details
+
+Yubikeys (Series 5 and on) have a built in smart card which provides PIV compatibility, which we can interact with using the [go-piv library](https://github.com/go-piv/piv-go).
+
+### yubikey login
+
+We'll add the `tsh` flag `--yubikey/-y` so that users can login with the new yubikey login flow. The majority of the login flow is the same, except the key generation and certificate signing process.
```

Comment:
Since the yubikey login can be required on the server auth settings, does the user need to provide the flag explicitly or can we detect that cluster/role requires it and enable this mode automatically?

Reply:
See https://github.com/gravitational/teleport/pull/14774#discussion_r932642276

Also, I think we want to give users the ability to do PIV login regardless of server settings.